src/typescript/azure-openapi-validator/rules/utilities/resourceUtils.ts

```diff
@@ -0,0 +1,507 @@
+        return [...pre, cur]
+      } else {
+        return pre
+      }
+    }, [])
+    return AllOfResources
+  }
+
+  public getAllOperationGetResponseModels() {
+    return this.getSpecificOperationModels("get", "200")
+  }
+
+  private getOperationGetResponseResources() {
+    const models = [...this.getAllOperationGetResponseModels().entries()].filter(m => this.checkResource(m[0]))
+    return new Map(models)
+  }
+
+  /**
+   * the result contains `put` resources
+   */
+  private getAllOperationsModels() {
+    const putOperationModels = unionModels(this.getSpecificOperationModels("put", "200"), this.getSpecificOperationModels("put", "201"))
+    const getOperationModels = this.getOperationGetResponseResources()
+    const operationModels = unionModels(putOperationModels, getOperationModels)
+    return operationModels
+  }
+
+  public getAllNestedResources() {
+    const fullModels = this.getOperationGetResponseResources()
+    const nestedModels = new Set<string>()
+    for (const entry of fullModels.entries()) {
+      const paths = entry[1]
+      paths
+        /* some paths not begin with /subscriptions/, they should be ignored */
+        .filter(p => p.toLowerCase().startsWith("/subscriptions/"))
+        .some(p => {
+          const hierarchy = this.getResourcesTypeHierarchy(p)
+          if (hierarchy.length > 1) {
+            nestedModels.add(entry[0])
+            return true
+          }
+        })
+    }
+    return nestedModels
+  }
+
+  /**
+   * Check the following conditions , a model be considered as a top-level resource
+   * 1 when a model existing in a get/put operation and 200/201 response, consider as a resource
+   * 2 when the path match the pattern: /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/expressRouteCircuits/{circuitName}
+   * 3 filter tenant wide resource
+   */
+  public getAllTopLevelResources() {
+    const fullModels = this.getOperationGetResponseResources()
+    const topLevelModels = new Set<string>()
+    for (const entry of fullModels.entries()) {
+      const paths = entry[1]
+      paths
+        .filter(p => p.toLowerCase().startsWith("/subscriptions/"))
+        .some(p => {
+          const hierarchy = this.getResourcesTypeHierarchy(p)
+          if (hierarchy.length === 1) {
+            topLevelModels.add(entry[0])
+            return true
+          }
+        })
+    }
+    return topLevelModels
+  }
+  /**
+   *
+   * @param path
+   * case 1 : '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/applicationGateways'
+   * return ["applicationGateways"]
+   * case 2: '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/expressRouteCircuits/{circuitName}/peerings/{peeringName}'
+   * return ["expressRouteCircuits","peerings"]
+   */
+  private getResourcesTypeHierarchy(path: string) {
```

Comment:
Do you have unit test for this function?

Reply:
private function